### ISSUE TITLE
ICE on Stable: failed to lift QueryResult

### ISSUE THREAD

**Issue by withkittens** [rust-lang/rust#52920]
Stable Rust 1.27.2 ICE'd with the following error:

```rust
error: internal compiler error: librustc/infer/canonical.rs:695: failed to lift `QueryResult { 
    var_values: CanonicalVarValues { var_values: ['?0, '?1] },
    region_constraints: [Binder(OutlivesPredicate(ReStatic, '?0))],
    certainty: Proven,
    value: NormalizationResult { normalized_ty: std::vec::Vec<usize> }
}`, canonicalized from `QueryResult { 
    var_values: CanonicalVarValues { var_values: ['_#0r, '_#1r] }, 
    region_constraints: [Binder(OutlivesPredicate(ReStatic, '_#0r))], 
    certainty: Proven,
    value: NormalizationResult { normalized_ty: _ }
}`
```

On Nightly compiles successfully.

**I tried the following code:**

[Playground](https://play.rust-lang.org/?gist=2f7d5bd6038ab10c936efdebddf92212&version=stable&mode=debug&edition=2015) (sorry, I can't make smaller bug-reproducible code snippet 🙁)

**Meta:**

* 1.27.2 (58cc626de 2018-07-18) running on x86_64-unknown-linux-gnu
* 1.29.0-nightly (54628c8ea 2018-07-30) running on x86_64-unknown-linux-gnu

<details>
<summary><strong>Backtrace:</strong></summary><p>

```
thread 'main' panicked at 'Box<Any>', librustc_errors/lib.rs:554:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:211
   3: std::panicking::default_hook
             at libstd/panicking.rs:227
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:467
   6: std::panicking::begin_panic
   7: rustc_errors::Handler::bug
   8: rustc::session::opt_span_bug_fmt::{{closure}}
   9: rustc::ty::context::tls::with_opt::{{closure}}
  10: rustc::ty::context::tls::with_context_opt
  11: rustc::ty::context::tls::with_opt
  12: rustc::session::opt_span_bug_fmt
  13: rustc::session::bug_fmt
  14: rustc::infer::canonical::Canonicalizer::canonicalize
  15: rustc_traits::util::make_query_response
  16: rustc::ty::context::tls::with_related_context
  17: rustc_traits::normalize_projection_ty::normalize_projection_ty
  18: rustc::ty::maps::<impl rustc::ty::maps::config::QueryConfig<'tcx> for rustc::ty::maps::queries::normalize_projection_ty<'tcx>>::compute
  19: rustc::dep_graph::graph::DepGraph::with_task_impl
  20: rustc::ty::context::tls::with_related_context
  21: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  22: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  23: <rustc::traits::query::normalize::QueryNormalizer<'cx, 'gcx, 'tcx> as rustc::ty::fold::TypeFolder<'gcx, 'tcx>>::fold_ty
  24: rustc::ty::fold::TypeFoldable::fold_with
  25: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &'a mut F>::call_once
  26: rustc::ty::fold::TypeFoldable::fold_with
  27: rustc::traits::query::normalize::<impl rustc::infer::at::At<'cx, 'gcx, 'tcx>>::normalize
  28: rustc::infer::InferCtxt::commit_if_ok
  29: rustc_mir::borrow_check::nll::type_check::TypeChecker::normalize
  30: <rustc_mir::borrow_check::nll::type_check::TypeVerifier<'a, 'b, 'gcx, 'tcx> as rustc::mir::visit::Visitor<'tcx>>::visit_constant
  31: <rustc_mir::borrow_check::nll::type_check::TypeVerifier<'a, 'b, 'gcx, 'tcx> as rustc::mir::visit::Visitor<'tcx>>::visit_mir
  32: rustc_mir::borrow_check::nll::type_check::type_check_internal
  33: rustc::ty::context::tls::with_related_context
  34: rustc::infer::InferCtxtBuilder::enter
  35: <rustc_mir::borrow_check::nll::type_check::TypeckMir as rustc_mir::transform::MirPass>::run_pass
  36: rustc_mir::transform::mir_const::{{closure}}
  37: rustc_mir::transform::mir_const
  38: rustc::ty::maps::<impl rustc::ty::maps::config::QueryConfig<'tcx> for rustc::ty::maps::queries::mir_const<'tcx>>::compute
  39: rustc::dep_graph::graph::DepGraph::with_task_impl
  40: rustc::ty::context::tls::with_related_context
  41: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  42: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  43: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::mir_const
  44: rustc_mir::transform::mir_validated
  45: rustc::ty::maps::<impl rustc::ty::maps::config::QueryConfig<'tcx> for rustc::ty::maps::queries::mir_validated<'tcx>>::compute
  46: rustc::dep_graph::graph::DepGraph::with_task_impl
  47: rustc::ty::context::tls::with_related_context
  48: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  49: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  50: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::ensure_query
  51: rustc_borrowck::borrowck::borrowck
  52: rustc::dep_graph::graph::DepGraph::with_task_impl
  53: rustc::ty::context::tls::with_related_context
  54: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  55: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  56: rustc_borrowck::borrowck::check_crate
  57: rustc::ty::context::tls::enter_context
  58: <std::thread::local::LocalKey<T>>::with
  59: rustc::ty::context::TyCtxt::create_and_enter
  60: rustc_driver::driver::compile_input
  61: rustc_driver::run_compiler_impl
  62: <scoped_tls::ScopedKey<T>>::set
  63: syntax::with_globals
  64: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  65: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:105
  66: rustc_driver::run
  67: rustc_driver::main
  68: std::rt::lang_start::{{closure}}
  69: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:310
  70: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:105
  71: std::rt::lang_start_internal
             at libstd/panicking.rs:289
             at libstd/panic.rs:374
             at libstd/rt.rs:58
  72: main
  73: __libc_start_main
  74: <unknown>
query stack during panic:
#0 [normalize_projection_ty] normalizing
    `Canonical { 
        variables: Slice([
            CanonicalVarInfo { kind: Region },
            CanonicalVarInfo { kind: Region }
        ]),
        value: ParamEnvAnd { 
            param_env: ParamEnv { 
                caller_bounds: Slice([]), 
                reveal: UserFacing
            },
            value: ProjectionTy { 
                substs: Slice([
                    std::iter::Map<impl std::iter::Iterator, [closure@src/main.rs:34:14: 44:17]>
                ]),
                item_def_id: DefId(2/0:1918 ~ core[cde8]::iter[0]::iterator[0]::Iterator[0]::Item[0])
            }
        }
    }`
#1 [mir_const] processing `find_saddle_points`
#2 [mir_validated] processing `find_saddle_points`
#3 [borrowck] processing `find_saddle_points`
end of query stack
error: aborting due to previous error
```
</p></details>

**Comment by estebank** [comment]
Seems like it's fixed already in 1.28, which lands soon.

**Comment by withkittens** [comment]
Yes, works on new stable. This may be closed then?
